### PR TITLE
Issue 4 and 5 support new version python

### PR DESCRIPTION
**Pull request by dk-WZFinTech**
Fix for issues: #4 and #5

### CODE REVIEW

File: sms_plusserver.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def __init__(self, response_text):
     def __repr__(self):
         return '<{} [{}]>'.format(self.__class__.__name__, self.message or '')
 
-    def __getitem__(self, item):
+    def __getitem__(self, item: Any):
```
Comment (piotrlukaj) [inline sms_plusserver.py]: I'd suggest to avoid using `Any`

File: sms_plusserver.py [hunk 2]
```diff
@@ -175,7 +182,12 @@ def configure(self, **kwargs):
 
     # High-level API:
 
-    def send(self, sms, timeout=None, fail_silently=False):
+    def send(
+        self,
+        sms: 'SMS',
+        timeout: Optional[float] = None,
```
Comment (piotrlukaj) [inline sms_plusserver.py:185]: Shouldn't it be of an `int` type? Regarding docs: `:param timeout: network timeout in seconds` it sounds like a solid `int` :)
(here and in other `timeouts` instances)

**Review by piotrlukaj — APPROVED** [review]
Added two questions, apart of that it looks fine 👍

File: sms_plusserver.py (hunk 2 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py:185]: ha... I also thought that on first place, but `sleep` can take values like: `0.5` and it is `float`. 

Secondary I go really deeper into `requests` docs (`requests/sessions.py:493`) and got this:
```python
class Session:

    def request():

        :param timeout: (optional) How long to wait for the server to send
            data before giving up, as a float, or a :ref:`(connect timeout,
            read timeout) <timeouts>` tuple.
```
that's why I used `float`

File: sms_plusserver.py (hunk 1 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: @escer any suggestion here?

File: .gitignore [hunk 3]
```diff
@@ -18,3 +18,7 @@ credentials.txt
 .eggs/
 dist/
 build/
+
+# DEV configs
```
Comment (escer) [inline .gitignore]: I'm undecided about these two. IMO we should adopt configs for flake8 and black and keep them in the repo.

File: sms_plusserver.py (hunk 1 above)
Comment (escer) [reply sms_plusserver.py]: @dk-WZFinTech I'd not use `Any` here as it does not provide *any* value. Let's skip type annotations when the only income it provides is more characters in the source code. It's a dynamic typed language after all.

File: sms_plusserver.py [hunk 4]
```diff
@@ -80,7 +80,7 @@ def __iter__(self):
     def items(self):
         return self._data.items()
 
-    def get(self, key, default=None):
+    def get(self, key: Any, default: Optional[Any] = None):
```
Comment (escer) [inline sms_plusserver.py]: I'd ignore type annotation here as well.

File: sms_plusserver.py [hunk 5]
```diff
@@ -155,7 +162,7 @@ def __repr__(self):
         return '<{} {}{}>'.format(
             self.__class__.__name__,
             self.username,
-            ' @ {}'.format(self.project) if self.project else ''
+            ' @ {}'.format(self.project) if self.project else '',
```
Comment (escer) [inline sms_plusserver.py]: BTW we can make it cleaner by using f-string here (but that's an optional suggestion).

File: sms_plusserver.py [hunk 6]
```diff
@@ -175,7 +182,12 @@ def configure(self, **kwargs):
 
     # High-level API:
 
-    def send(self, sms, timeout=None, fail_silently=False):
+    def send(
+        self,
+        sms: 'SMS',
+        timeout: Optional[float] = None,
+        fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]: `fail_silently: bool = False,`

File: sms_plusserver.py [hunk 7]
```diff
@@ -175,7 +182,12 @@ def configure(self, **kwargs):
 
     # High-level API:
 
-    def send(self, sms, timeout=None, fail_silently=False):
+    def send(
+        self,
+        sms: 'SMS',
```
Comment (escer) [inline sms_plusserver.py:184]: Let's add `from __future__ import annotations` and use `sms: SMS` here.

File: sms_plusserver.py [hunk 8]
```diff
@@ -214,13 +226,19 @@ def send(self, sms, timeout=None, fail_silently=False):
             result = bool(put_response and put_response.is_ok)
         return result
 
-    def check_state(self, sms, wait=False, timeout=None, fail_silently=False):
+    def check_state(
+        self,
+        sms: 'SMS',
```
Comment (escer) [inline sms_plusserver.py:228]: Let's add `from __future__ import annotations` and use `sms: SMS` here.

File: sms_plusserver.py [hunk 9]
```diff
@@ -214,13 +226,19 @@ def send(self, sms, timeout=None, fail_silently=False):
...
+        wait: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        wait: bool = False,
```

File: sms_plusserver.py [hunk 10]
```diff
@@ -214,13 +226,19 @@ def send(self, sms, timeout=None, fail_silently=False):
...
+        timeout: Optional[float] = None,
+        fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 11]
```diff
@@ -249,10 +267,18 @@ def check_state(self, sms, wait=False, timeout=None, fail_silently=False):
...
     def put_sms(
-        self, destination, text, orig=None, registered_delivery=True,
-        debug=False, project=None, encoding=None, max_parts=None, timeout=None,
-        fail_silently=False
-    ):
+        self,
+        destination: str,
+        text: str,
+        orig: Optional[str] = None,
+        registered_delivery: Optional[bool] = True,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        registered_delivery: bool = True,
```

File: sms_plusserver.py [hunk 12]
```diff
@@ -249,10 +267,18 @@ def check_state(self, sms, wait=False, timeout=None, fail_silently=False):
...
+        debug: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        debug: bool = False,
```

File: sms_plusserver.py [hunk 13]
```diff
@@ -249,10 +267,18 @@ def check_state(self, sms, wait=False, timeout=None, fail_silently=False):
...
+        project: Optional[str] = None,
+        encoding: Optional[str] = None,
+        max_parts: Optional[int] = None,
+        timeout: Optional[float] = None,
+        fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 14]
```diff
@@ -299,17 +325,26 @@ def put_sms(
...
+            resource_name='Put SMS',
+            fail_silently=fail_silently,
         )
 
-    def check_sms_state(self, handle_id, timeout=None, fail_silently=False):
+    def check_sms_state(
+        self,
+        handle_id: str,
+        timeout: Optional[float] = None,
+        fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 15]
```diff
@@ -339,12 +374,21 @@ def check_sms_state(self, handle_id, timeout=None, fail_silently=False):
...
+            fail_silently=fail_silently,
         )
 
-    def wait_until_arrived(self, handle_id, timeout=None, fail_silently=False):
-        """Waits until SMS (Handle ID) gets "arrived" state on Plusserver SMS
+    def wait_until_arrived(
+        self,
+        handle_id: str,
+        timeout: Optional[float] = None,
+        fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 16]
```diff
@@ -453,8 +503,15 @@ class SMS(object):
     """Single message wrapper"""
 
     def __init__(
-        self, destination, text, orig=None, registered_delivery=True,
-        debug=False, project=None, encoding=None, max_parts=None
+        self,
+        destination: str,
+        text: str,
+        orig: Optional[str] = None,
+        registered_delivery: Optional[bool] = True,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        registered_delivery: bool = True,
```

File: sms_plusserver.py [hunk 17]
```diff
@@ -453,8 +503,15 @@ class SMS(object):
...
+        debug: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        debug: bool = False,
```

File: sms_plusserver.py [hunk 18]
```diff
@@ -491,10 +548,15 @@ def __repr__(self):
...
             ' [{}]'.format(handle_id) if handle_id else '',
-            ' {}'.format(state) if state else ''
+            ' {}'.format(state) if state else '',
         )
 
-    def send(self, timeout=None, fail_silently=False, service=None):
+    def send(
+        self,
+        timeout: Optional[float] = None,
+        fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 19]
```diff
@@ -512,11 +574,15 @@ def send(self, timeout=None, fail_silently=False, service=None):
         return service.send(self, timeout=timeout, fail_silently=fail_silently)
 
     def check_state(
-        self, wait=False, timeout=None, fail_silently=False, service=None
-    ):
+        self,
+        wait: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        wait: bool = False,
```

File: sms_plusserver.py [hunk 20]
```diff
@@ -512,11 +574,15 @@ def send(self, timeout=None, fail_silently=False, service=None):
...
+        timeout: Optional[float] = None,
+        fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
        fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 21]
```diff
@@ -535,24 +601,33 @@ def check_state(
...
+
 def send_sms(
-    destination, text, orig=None, registered_delivery=True,
-    debug=False, project=None, encoding=None, max_parts=None, timeout=None,
-    fail_silently=False, service=None
-):
+    destination: str,
+    text: str,
+    orig: Optional[str] = None,
+    registered_delivery: Optional[bool] = True,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
    registered_delivery: bool = True,
```

File: sms_plusserver.py [hunk 22]
```diff
@@ -535,24 +601,33 @@ def check_state(
...
+    debug: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
    debug: bool = False,
```

File: sms_plusserver.py [hunk 23]
```diff
@@ -535,24 +601,33 @@ def check_state(
...
+    project: Optional[str] = None,
+    encoding: Optional[str] = None,
+    max_parts: Optional[int] = None,
+    timeout: Optional[float] = None,
+    fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
    fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 24]
```diff
@@ -581,17 +656,26 @@ def send_sms(
...
         timeout=timeout, fail_silently=fail_silently, service=service
     )
 
 
-def check_sms_state(handle_id, timeout=None, fail_silently=False,
-                    service=None):
+def check_sms_state(
+    handle_id: str,
+    timeout: Optional[float] = None,
+    fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
    fail_silently: bool = False,
```

File: sms_plusserver.py [hunk 25]
```diff
@@ -615,9 +699,12 @@ def check_sms_state(handle_id, timeout=None, fail_silently=False,
 
 
 def wait_until_arrived(
-    handle_id, timeout=None, fail_silently=False, service=None
-):
-    """Waits until SMS gets "arrived" status.
+    handle_id: str,
+    timeout: Optional[float] = None,
+    fail_silently: Optional[bool] = False,
```
Comment (escer) [inline sms_plusserver.py]:
```suggestion
    fail_silently: bool = False,
```

File: tests.py [hunk 26]
```diff
@@ -546,11 +535,9 @@ def test_check_sms_state_error_network(self, mock_post):
 
         mock_post.assert_called_once_with(
             sms_plusserver.SMSService.SMS_STATE_URL,
-            {
-                'handle': 'd41d8cd98f00b204e9800998ecf8427e',
-            },
+            {'handle': 'd41d8cd98f00b204e9800998ecf8427e',},
```
Comment (escer) [inline tests.py]: Please reconfigure your Black so it wouldn't leave the trailing commas on single-line literals.

**Review by escer — CHANGES_REQUESTED** [review]
Great work with switching the platform from Python 2.7 to Python 3.8, but there are still some places where fixes are required.

File: sms_plusserver.py (hunk 7 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py:184]: @escer `from __future__ import annotations` is supported from python 3.7 and up. If we will use it then we need to drop support for python 3.6 as well. What you think?

File: sms_plusserver.py (hunk 9 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done

File: sms_plusserver.py (hunk 10 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done

File: sms_plusserver.py (hunk 11 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done

File: sms_plusserver.py (hunk 6 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 22 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 23 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 24 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 25 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 5 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/cab18898f97c32d5a8ed5e49a81deddf509b5ac7

File: sms_plusserver.py (hunk 1 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: - [x] I will remove it - then

File: .gitignore (hunk 3 above)
Comment (dk-WZFinTech) [reply .gitignore]: - [x] I will remove it

File: tests.py (hunk 26 above)
Comment (dk-WZFinTech) [reply tests.py]: - [x] I will remove  trailing commas

File: sms_plusserver.py (hunk 1 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/548f9fde0d8759ba95fcb55ba2c417194e39dd2f

File: sms_plusserver.py (hunk 4 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/548f9fde0d8759ba95fcb55ba2c417194e39dd2f

File: sms_plusserver.py (hunk 12 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 13 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 14 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 15 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 16 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 17 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 19 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 18 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 20 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: sms_plusserver.py (hunk 21 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/f11e24bd49fe2df4d0be8e4beb85316dc967fb04

File: .gitignore (hunk 3 above)
Comment (dk-WZFinTech) [reply .gitignore]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/528d94412bf4084b9bd10b6130d491dd187e033a

File: sms_plusserver.py (hunk 7 above)
Comment (escer) [reply sms_plusserver.py:184]: Let's leave it as it is then.

Comment (dk-WZFinTech) [reply sms_plusserver.py:184]: As we discussed I will skip this solution.

File: sms_plusserver.py (hunk 8 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py:228]: Same as here:
https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6#discussion_r670384862

File: tests.py (hunk 26 above)
Comment (dk-WZFinTech) [reply tests.py]: Done with https://github.com/W-Z-FinTech-GmbH/sms_plusserver/pull/6/commits/3f724767adb145aac9d5fcc842fea765ed9d774b

**Comment by dk-WZFinTech** [comment]
@escer I implemented all required changes - please re-review it.

**Review by escer — APPROVED** [review]

File: sms_plusserver.py [hunk 27]
```diff
@@ -299,17 +322,26 @@ def put_sms(
...
+            timeout=timeout,
+            resource_name='Put SMS',
+            fail_silently=fail_silently,
         )
 
-    def check_sms_state(self, handle_id, timeout=None, fail_silently=False):
+    def check_sms_state(
+        self,
+        handle_id: str,
+        timeout: Optional[float] = None,
```
Comment (piotrlukaj) [inline sms_plusserver.py:342]: I'd keep the `Optional` parameters at the end, before the required ones. (or the `fail_silently` is Optional and missing the typehint?)

Here and in the methods below.

**Review by piotrlukaj — APPROVED** [review]
One nitpick regarding the typehints, apart of that it looks good!

File: sms_plusserver.py (hunk 27 above)
Comment (dk-WZFinTech) [reply sms_plusserver.py:342]: > ... the `fail_silently` is Optional

Yes, it is but `Optional` is more like `Union` and it is used only to combine `None` and other type.

**Pull request merged**